Add more options in findAll

AnthonyAmanse opened the pull request:
This commit adds in options:
* ORDER BY clause for sorting
* OFFSET and LIMIT for retrieving a portion of the rows

This builds the appropriate SQL queries for the options listed above.
This commit also adds in initial tests for the new options.

Signed-off-by: AnthonyAmanse <ghieamanse@gmail.com>

EnriqueL8 commented:
Hey @AnthonyAmanse, this looks really good! Just a few comments: 
 - We have some work planned around `QueryParams` and filtering so it would be good to just have these options in the `findAll` calls without `QueryParams`
- Some verification needs to take place with the `OrderBy` parameter to prevent SQL Injection - for example verifying that the column is actually in the table? or using the `Parameter()` function from SwiftKuery

EnriqueL8 commented:
One more suggestion: would be to make the `orderBy` parameter accept multiple `OrderBy` instance since for example we would want to order by ascending name and descending age. The definition could be: `orderBy: OrderBy? ...`

AnthonyAmanse committed:
Add more options in findAll

This commit adds in options:
* ORDER BY clause for sorting
* OFFSET and LIMIT for retrieving a portion of the rows

This builds the appropriate SQL queries for the options listed above.
This commit also adds in initial tests for the new options.

Signed-off-by: AnthonyAmanse <ghieamanse@gmail.com>

AnthonyAmanse commented on Tests/SwiftKueryORMTests/TestFind.swift:
```diff
@@ -113,4 +118,139 @@ class TestFind: XCTestCase {
+        var column: Column?
+        do {
+            let table = try Person.getTable()
+            column = table.columns.first(where: {$0.name == "age"})
```
@EnriqueL8 Would this be enough for a verification if the column exists? Since the `orderBy` parameter is an OrderBy buildable that doesn't accept just any string, verification would have to be done outside of `findAll` function

EnriqueL8 replied:
Annoyingly yes - the `OrderBy` only accepts a `Field` type that can be for example a Column but also a [RawField](https://github.com/IBM-Swift/Swift-Kuery/blob/master/Sources/SwiftKuery/RawField.swift). 
We could use [Order](https://github.com/IBM-Swift/KituraContracts/blob/master/Sources/KituraContracts/Contracts.swift#L898) that accepts a *String* and then find a *Column* matching that *String* and then create the `OrderBy` case.

AnthonyAmanse reviewed (commented):
@EnriqueL8 Thanks for the feedback. Updated the function to accept a variadic parameter of `orderBy`. Not sure what to do with verification for columns though

EnriqueL8 commented:
@AnthonyAmanse Sorry - this needs now a rebase since your Swift-Kuery 2.0.0 PR just got merged and addressing the comments

EnriqueL8 commented on Sources/SwiftKueryORM/Model.swift:
```diff
@@ -992,12 +1014,34 @@ public extension Model {
-    let query = Select(from: table)
+    var query = Select(from: table)
+    var orderByArray: [OrderBy] = []
+    for order in order {
```
Could you extract this logic into a function?

AnthonyAmanse commented:
@EnriqueL8 Moved the logic to `getOrderBy`

EnriqueL8 commented on Sources/SwiftKueryORM/Model.swift:
```diff
@@ -919,6 +919,25 @@ public extension Model {
+  /**
+    This function constructs an array of OrderBy from the Order values
+  */
+  private static func getOrderBy(order: [Order], table: Table) throws -> [OrderBy] {
```
This method looks quite familiar to the above method, maybe we could call this method from the above to remove code duplication

EnriqueL8 commented on Sources/SwiftKueryORM/Model.swift:
```diff
@@ -919,6 +919,25 @@ public extension Model {
+    var orderByArray: [OrderBy] = []
+    for order in order {
+      guard let column = table.columns.first(where: {$0.name == order.value}) else {
+        throw RequestError(.ormInternalError, reason: "Column \(order.value) not found")
```
Instead of throwing and `.ormInternalError`, we should throw `.ormQueryError` because it failed during the construction of the Query?

EnriqueL8 commented on Sources/SwiftKueryORM/Model.swift:
```diff
@@ -919,6 +919,25 @@ public extension Model {
+      guard let column = table.columns.first(where: {$0.name == order.value}) else {
+        throw RequestError(.ormInternalError, reason: "Column \(order.value) not found")
+      }
+      if case .asc(order.value) = order {
```
Can we remove the `order.value`? And just have `.asc(_)`?

EnriqueL8 commented on Tests/SwiftKueryORMTests/CommonUtils.swift:
```diff
@@ -136,12 +142,19 @@ class TestConnection: Connection {
-    init(numberOfRows: Int) {
+    init(numberOfRows: Int, sortedByAge: String? = nil) {
         self.numberOfRows = numberOfRows
+        if let sortedByAge = sortedByAge {
```
Could we sort the array programatically? By using `array.sort`? In the case of having to change the rows we would also have to update these two lines.

EnriqueL8 reviewed (commented):
Sorry - I promise these are the last changes that I suggest!

AnthonyAmanse commented:
@EnriqueL8 Feedbacks are always great! Merged the `getOrderBy` function instead of overloading. Sorted the `rows` using `rows.sort` instead of hardcoding it.

EnriqueL8 commented:
LGTM - Updated Branch because it is out-of-date

djones6 committed:
Merge branch 'master' into additional-options
